op_crates/crypto/lib.rs

```diff
@@ -43,6 +81,441 @@ pub fn op_crypto_get_random_values(
 }
 
+struct CryptoKeyResource<A> {
+  crypto_key: WebCryptoKey,
+  key: A,
+  hash: Option<WebCryptoHash>,
+}
+
+impl Resource for CryptoKeyResource<RSAPublicKey> {
+  fn name(&self) -> Cow<str> {
+    "RSAPublicCryptoKey".into()
+  }
+}
+
+impl Resource for CryptoKeyResource<RSAPrivateKey> {
+  fn name(&self) -> Cow<str> {
+    "RSAPrivateCryptoKey".into()
+  }
+}
+
+impl Resource for CryptoKeyResource<EcdsaKeyPair> {
+  fn name(&self) -> Cow<str> {
+    "ECDSACryptoKey".into()
+  }
+}
+
+impl Resource for CryptoKeyResource<ring::agreement::PublicKey> {
+  fn name(&self) -> Cow<str> {
+    "ECDHPublicKey".into()
+  }
+}
+
+impl Resource for CryptoKeyResource<EphemeralPrivateKey> {
+  fn name(&self) -> Cow<str> {
+    "ECDHPrivateKey".into()
+  }
+}
+
+impl Resource for CryptoKeyResource<HmacKey> {
+  fn name(&self) -> Cow<str> {
+    "HMACKey".into()
+  }
+}
+
+#[derive(Deserialize)]
+#[serde(rename_all = "camelCase")]
+struct WebCryptoAlgorithmArg {
+  name: Algorithm,
+  modulus_length: Option<u32>,
+  hash: Option<WebCryptoHash>,
+  #[allow(dead_code)]
+  length: Option<u32>,
+  named_curve: Option<WebCryptoNamedCurve>,
+}
+
+#[derive(Deserialize)]
+#[serde(rename_all = "camelCase")]
+struct WebCryptoGenerateKeyArg {
+  algorithm: WebCryptoAlgorithmArg,
+  extractable: bool,
+  key_usages: Vec<KeyUsage>,
+}
+
+#[derive(Serialize)]
+#[serde(rename_all = "camelCase")]
+enum JSCryptoKey {
+  Single {
+    key: WebCryptoKey,
+    rid: u32,
+  },
+  Pair {
+    key: WebCryptoKeyPair,
+    private_rid: u32,
+    public_rid: u32,
+  },
+}
+
+macro_rules! validate_usage {
+  ($e: expr, $u: expr) => {
+    for usage in $e {
+      if !$u.contains(&usage) {
+        return Ok(json!({ "err": DOMError("Invalid key usage".to_string()) }))
+      }
+    }
+  }
+}
+
+pub async fn op_webcrypto_generate_key(
+  state: Rc<RefCell<OpState>>,
+  args: Value,
+  zero_copy: BufVec,
+) -> Result<Value, AnyError> {
+  let args: WebCryptoGenerateKeyArg = serde_json::from_value(args)?;
+  let extractable = args.extractable;
+  let algorithm = args.algorithm.name;
+
+  let mut state = state.borrow_mut();
+
+  let key = match algorithm {
+    Algorithm::RsassaPkcs1v15 | Algorithm::RsaPss => {
+      validate_usage!(&args.key_usages, vec![KeyUsage::Sign, KeyUsage::Verify]);
+
+      let public_exponent = if !zero_copy.is_empty() {
+        Some(&*zero_copy[0])
+      } else {
+        None
+      };
+
+      let exp = public_exponent.ok_or_else(|| {
+        WebCryptoError::MissingArgument("publicExponent".to_string())
+      })?;
+      let modulus_length = args.algorithm.modulus_length.ok_or_else(|| {
+        WebCryptoError::MissingArgument("modulusLength".to_string())
+      })?;
+
+      let exponent = BigUint::from_bytes_be(exp);
+
+      // Generate RSA private key based of exponent, bits and Rng.
+      let mut rng = OsRng;
+      let private_key = RSAPrivateKey::new_with_exp(
+        &mut rng,
+        modulus_length as usize,
+        &exponent,
+      )?;
```

Comment:
I'm guessing this is a CPU intensive function (if not this one exactly then probably some other call in this op). This op is asynchronous, but the way it's organized now this CPU-heavy function is going to block the event loop.

I think this whole bit should be wrapped in an async block and passed to `tokio::task::spawn_blocking()`.

See also https://docs.rs/tokio/1.2.0/tokio/#cpu-bound-tasks-and-blocking-code

Reply:
done :+1: